nodejs/.jest/setup.js

```diff
@@ -0,0 +1,87 @@
+// Here we define some mock functions and some globally accessible utility functions that get used across different tests.
+
+/* -------------------------------------------- Globally mocked functions ------------------------------------------- */
+
+// Replace every instance of console.log and console.error with a mock function. This way, we can keep track of the number of times they are called
+// and replace their functionality with a simple mock (that does nothing) so that we don't actually log anything to the console during testing.
+// TODO: Update this when we find a better way to do errors (ideally we should just be raising an exception here and logging elsewhere).
+global.console = {
+  log: jest.fn(),
+
+  // Keep native behaviour for other methods, use those to print out things in your own tests
+  error: console.error,
+  warn: console.warn,
+  info: console.info,
+  debug: console.debug,
+};
+
+// Mock for the client.views.open method
+global.openViewMockFunc = jest.fn();
+// Mock for the client.views.publish method
+global.publishViewMockFunc = jest.fn();
+// Mocks the client.chat.update method
+global.updateChatMockFunc = jest.fn();
+// Mock for the ack() method
+global.ackMockFunc = jest.fn();
+
+/* --------------------------------------------- Global helper functions -------------------------------------------- */
+
+// A helper function to parse and validate JSON.
+global.isValidJSON = (jsonString) => {
+  try {
+    var obj = JSON.parse(jsonString);
+
+    // Handle non-exception-throwing cases:
+    // Neither JSON.parse(false) or JSON.parse(1234) throw errors, hence the type-checking,
+    // but... JSON.parse(null) returns null, and typeof null === "object",
+    // so we must check to make sure the object isn't null as well.
+    // Thankfully, null is falsy, so this suffices:
+    if (obj && typeof obj === 'object') {
+      return obj;
+    }
+  } catch (e) {}
+
+  return false;
+};
+
+global.testListener = async (
+  callbackFunctionPromiseToTest,
+  mockedApiMethod,
+  mockedApiMethodArgObj,
```

Comment:
That's a great point! I was thinking of the first point but I wasn't sure if it was better to have the function only accept one pair of `apiMethod: args` and just have separate tests individually for each. That way, the unit test failing will explicitly tell you which method isn't being called 🤔 . I'll try it out and see what it's like.

I do like the second implementation you mentioned with the callback functions, it's less abstraction but also a lot more flexible. It will cause a bit more code repetition across the tests since the assertions are very similar, but I think that will be worth the flexibility. I don't want to overabsract.

Reply:
I agree with all in your comment here. Look forward to seeing the updated one!